pyxem/signals/segments.py

```diff
@@ -0,0 +1,483 @@
+            LearningSegment where the loadings have been segmented and some
+            factors have been repeated according to the new number of loading
+            segments.
+        """
+
+        factors = self.factors.copy()
+        loadings = self.loadings.copy()
+        loadings_shape_x = loadings.data.shape[1]
+        loadings_shape_y = loadings.data.shape[2]
+        factors_shape_x = factors.data.shape[1]
+        factors_shape_y = factors.data.shape[2]
+
+        loadings_segments = np.array(loadings.map(
+            separate_watershed, show_progressbar=True, inplace=False,
+            min_distance=min_distance, min_size=min_size, max_size=max_size,
+            max_number_of_grains=max_number_of_grains,
+            marker_radius=marker_radius, threshold=threshold,
+            exclude_border=exclude_border), dtype=np.object)
+
+        segments, factors_of_segments = [], []
+        num_segs_tot = 0
+        for loading_segment, factor in zip(loadings_segments, factors):
+            segments = np.append(segments, loading_segment)
+            num_segs = np.shape(loading_segment)[0]
+            factors_of_segments = np.append(
+                factors_of_segments, np.broadcast_to(
+                    factor, (num_segs, factors_shape_x, factors_shape_y)))
+            num_segs_tot += num_segs
+        segments = segments.reshape(
+            (num_segs_tot, loadings_shape_y, loadings_shape_x))
+        factors_of_segments = factors_of_segments.reshape(
+            (num_segs_tot, factors_shape_x, factors_shape_y))
+
+        delete_indices = list(map(lambda x: x.max() < min_intensity_threshold, segments))
+        delete_indices = np.where(delete_indices)
+        segments = np.delete(segments, delete_indices, axis=0)
+        factors_of_segments = np.delete(
+            factors_of_segments, delete_indices, axis=0)
+
+        # if TraitError is raised, it is likely no segements were found
+        segments = Signal2D(segments).transpose(navigation_axes=[0],
+                                                signal_axes=[2, 1])
+        factors_of_segments = Signal2D(factors_of_segments)
+        learning_segment = LearningSegment(segments, factors_of_segments)
+        return learning_segment
+
+
+class VDFSegment:
+    _signal_type = "vdf_segment"
+
+    def __init__(self, segments, vectors_of_segments, intensities=None):
+        # Segments as Signal2D
+        self.segments = segments
+        # DiffractionVectors
+        self.vectors_of_segments = vectors_of_segments
+        # Intensities corresponding to each vector
+        self.intensities = intensities
+
+    def get_ncc_matrix(self):
+        """Get the normalised correlation coefficient (NCC) matrix containing
+        the NCC between each pair of segments.
+
+        Returns
+        -------
+        ncc_matrix : Signal2D
+            Normalised correlation coefficient matrix.
+        """
+        # Set up an empty matrix of correct size to store NCC values.
```

Comment:
We should probably factor out the code here so it can be re-used in the LearningSegment equivalent function.

Reply:
will leave this as a minor point for future and add a TODO in the code.